ffi/src/ir.rs

```diff
@@ -0,0 +1,326 @@
+        Self(source)
+    }
+}
+
+impl<T: Ir> From<Vec<T>> for IrVec<T, T::Type> {
+    fn from(source: Vec<T>) -> Self {
+        Self(source)
+    }
+}
+
+impl<T: Transmute> Transparent<T> {
+    /// Convert from [`Self`] into [`T::Inner`].
+    /// Type returned is guaranteed to be valid
+    pub(crate) fn into_inner(self) -> T::Target {
+        #[repr(C)]
+        union TransmuteHelper<T: Transmute> {
+            source: ManuallyDrop<T>,
+            target: ManuallyDrop<T::Target>,
+        }
+
+        let transmute_helper = TransmuteHelper {
+            source: ManuallyDrop::new(self.0),
+        };
+
+        // SAFETY: Transmute is always valid because T::ReprC is a superset of T
+        ManuallyDrop::into_inner(unsafe { transmute_helper.target })
+    }
+
+    /// Convert from [`T::Inner`] into [`Self`]
+    ///
+    /// # Safety
+    ///
+    /// Any raw pointer in [`Self::Inner`] that will be dereferenced must be valid.
+    pub(crate) unsafe fn try_from_inner(source: T::Target) -> Option<Self> {
+        #[repr(C)]
+        union TransmuteHelper<T: Transmute> {
+            source: ManuallyDrop<T::Target>,
+            target: ManuallyDrop<T>,
+        }
+
+        if !T::is_valid(&source) {
+            return None;
+        }
+
+        let transmute_helper = TransmuteHelper {
+            source: ManuallyDrop::new(source),
+        };
+        Some(Transparent(ManuallyDrop::into_inner(
+            transmute_helper.target,
+        )))
+    }
+}
+
+// SAFETY: Trivially transmutable
+unsafe impl<T: Transmute> Transmute for &T {
+    type Target = *const T::Target;
+
+    #[inline]
+    unsafe fn is_valid(source: &*const T::Target) -> bool {
+        if let Some(item) = source.as_ref() {
+            return T::is_valid(item);
+        }
+
+        false
+    }
+}
+
+// SAFETY: Trivially transmutable
+unsafe impl<T: Transmute> Transmute for &mut T {
+    type Target = *mut T::Target;
+
+    #[inline]
+    unsafe fn is_valid(source: &*mut T::Target) -> bool {
+        if let Some(item) = source.as_ref() {
+            return T::is_valid(item);
+        }
+
+        false
+    }
+}
+
+impl<'itm, T: Ir + NonTransmute> Ir for &'itm T {
+    type Type = &'itm T;
+}
+
+impl<'itm, T> Ir for &'itm mut T
+where
+    &'itm mut T: Transmute,
+{
+    type Type = Transparent<&'itm mut T>;
+}
+
+impl<T> Ir for *const T {
+    type Type = AsIs<Self>;
+}
+
+impl<T> Ir for *mut T {
+    type Type = AsIs<Self>;
+}
+
+impl<T: Ir> Ir for Box<T> {
+    type Type = IrBox<T, T::Type>;
+}
+impl<'itm, T: Ir> Ir for &'itm [T] {
+    type Type = IrSlice<'itm, T, T::Type>;
+}
+impl<'itm, T: Ir> Ir for &'itm mut [T] {
+    type Type = IrSliceMut<'itm, T, T::Type>;
+}
+impl<T: Ir> Ir for Vec<T> {
+    type Type = IrVec<T, T::Type>;
+}
+
+// NOTE: This can be contested as it is nowhere documented that String is
+// actually transmutable into Vec<u8>, but implicitly it should be
+// SAFETY: Vec<String> type should be transmutable into Vec<u8>
```

Comment:
Yes, as far as I was able to find out it is a fair assumption that `String` is a wrapper for `Vec<u8>`. Explicitly it isn't transparent, but implicitly it should be. This conclusion stems from the API of a `String`. This is something to have in mind so I left a note. We should reconsider the String conversion at some point to rely only on guaranteed representations

Reply:
I didn't want to worry about special handling `String` impls when the assumption that it is transmutable into `Vec<u8>` seems quite reasonable. This can be dealt with in a separate issue if deemed necessary

Reply:
We shouldn't rely on API features of the standard lib. Given that `String` is given no special treatment syntactically, we could very well suggest people just use `ConstString` and nothing else.

Reply:
ok, I opened #2795